File: go/store_test.go
package featurevault

import "testing"

// staleness returns a pointer to a max-staleness bound for the optional argument.
func staleness(n int64) *int64 { return &n }

func newStore() *FeatureStore {
	s := NewFeatureStore()
	for _, o := range []observation{{100, 1.0}, {200, 2.0}, {300, 3.0}} {
		s.Ingest(FeatureValue{"user-1", o.t, o.v})
	}
	return s
}

func mustValue(t *testing.T, got *float64, want float64) {
	t.Helper()
	if got == nil {
		t.Fatalf("expected %v, got nil", want)
	}
	if *got != want {
		t.Fatalf("expected %v, got %v", want, *got)
	}
}

func mustNil(t *testing.T, got *float64) {
	t.Helper()
	if got != nil {
		t.Fatalf("expected nil, got %v", *got)
	}
}

func TestAsOfJoinPicksLatestBeforeEvent(t *testing.T) {
	s := newStore()
	mustValue(t, s.GetPointInTime("user-1", 250, nil), 2.0) // latest <= 250 is ts=200
	mustValue(t, s.GetPointInTime("user-1", 300, nil), 3.0) // inclusive of event time
	mustValue(t, s.GetPointInTime("user-1", 100, nil), 1.0)
}

func TestNoLeakageBeforeFirstValue(t *testing.T) {
	mustNil(t, newStore().GetPointInTime("user-1", 50, nil))
}

func TestNoFutureLeak(t *testing.T) {
	// At event time 150 only the ts=100 value existed; ts=200/300 are the future.
	mustValue(t, newStore().GetPointInTime("user-1", 150, nil), 1.0)
}

func TestMaxStalenessExpiresOldFeatures(t *testing.T) {
	s := newStore()
	// Newest value at event 1000 is ts=300 -> 700 old; with staleness 100 it expires.
	mustNil(t, s.GetPointInTime("user-1", 1000, staleness(100)))
	mustValue(t, s.GetPointInTime("user-1", 350, staleness(100)), 3.0)
}

func TestMaxStalenessBoundaryIsInclusive(t *testing.T) {
	s := newStore()
	// event 400, ts=300 -> exactly 100 old; staleness 100 is not "> 100", so kept.
	mustValue(t, s.GetPointInTime("user-1", 400, staleness(100)), 3.0)
	// One second older tips it over the boundary and expires it.
	mustNil(t, s.GetPointInTime("user-1", 401, staleness(100)))
}

func TestUnknownEntityReturnsNone(t *testing.T) {
	s := newStore()
	mustNil(t, s.GetPointInTime("ghost", 200, nil))
	mustNil(t, s.GetOnline("ghost"))
}

func TestTrainingSetIsPointInTimeCorrect(t *testing.T) {
	s := newStore()
	spine := []SpineRow{{"user-1", 120}, {"user-1", 220}, {"user-1", 320}}
	col := s.GetTrainingSet(spine, nil)
	want := []float64{1.0, 2.0, 3.0}
	if len(col) != len(want) {
		t.Fatalf("length mismatch: got %d want %d", len(col), len(want))
	}
	for i, w := range want {
		mustValue(t, col[i], w)
	}
}

func TestTrainingSetLeavesGapsAsNil(t *testing.T) {
	s := newStore()
	// First row precedes any value; unknown entity has no value either.
	spine := []SpineRow{{"user-1", 50}, {"user-1", 250}, {"ghost", 999}}
	col := s.GetTrainingSet(spine, nil)
	mustNil(t, col[0])
	mustValue(t, col[1], 2.0)
	mustNil(t, col[2])
}

func TestOnlineReturnsLatest(t *testing.T) {
	mustValue(t, newStore().GetOnline("user-1"), 3.0)
}

func TestOutOfOrderIngestIsHandled(t *testing.T) {
	s := NewFeatureStore()
	for _, o := range []observation{{300, 3.0}, {100, 1.0}, {200, 2.0}} {
		s.Ingest(FeatureValue{"e", o.t, o.v})
	}
	mustValue(t, s.GetPointInTime("e", 250, nil), 2.0)
	mustValue(t, s.GetOnline("e"), 3.0)
}

func TestDuplicateTimestampLargestValueWins(t *testing.T) {
	// The documented tie-break: values are ordered by (timestamp, value), so
	// among equal timestamps the largest value is served.
	s := NewFeatureStore()
	for _, o := range []observation{{200, 5.0}, {200, 2.0}, {200, 9.0}, {100, 1.0}} {
		s.Ingest(FeatureValue{"e", o.t, o.v})
	}
	mustValue(t, s.GetPointInTime("e", 250, nil), 9.0)
	mustValue(t, s.GetOnline("e"), 9.0)
	// A pick at ts=100 is unaffected by the later duplicates.
	mustValue(t, s.GetPointInTime("e", 150, nil), 1.0)
}

func TestEmptyStoreReturnsNil(t *testing.T) {
	s := NewFeatureStore()
	mustNil(t, s.GetPointInTime("nobody", 10, nil))
	mustNil(t, s.GetOnline("nobody"))
	if col := s.GetTrainingSet(nil, nil); len(col) != 0 {
		t.Fatalf("expected empty column, got %v", col)
	}
}

func TestSingleObservationBoundaries(t *testing.T) {
	s := NewFeatureStore()
	s.Ingest(FeatureValue{"e", 100, 1.0})
	mustNil(t, s.GetPointInTime("e", 99, nil)) // strictly before -> nothing
	mustValue(t, s.GetPointInTime("e", 100, nil), 1.0)
	mustValue(t, s.GetPointInTime("e", 101, nil), 1.0)
}

func TestNegativeAndZeroValuesAreServed(t *testing.T) {
	s := NewFeatureStore()
	s.Ingest(FeatureValue{"e", 10, 0.0})
	s.Ingest(FeatureValue{"e", 20, -3.5})
	mustValue(t, s.GetPointInTime("e", 15, nil), 0.0)
	mustValue(t, s.GetPointInTime("e", 25, nil), -3.5)
	mustValue(t, s.GetOnline("e"), -3.5)
}

File: ts/src/store.test.ts
import { describe, expect, it } from "vitest";

import { FeatureStore } from "./store.ts";

function store(): FeatureStore {
  const s = new FeatureStore();
  for (const [t, v] of [
    [100, 1.0],
    [200, 2.0],
    [300, 3.0],
  ] as const) {
    s.ingest({ entity: "user-1", timestamp: t, value: v });
  }
  return s;
}

describe("FeatureStore as-of join", () => {
  it("picks the latest value at or before the event time", () => {
    const s = store();
    expect(s.getPointInTime("user-1", 250)).toBe(2.0); // latest <= 250 is ts=200
    expect(s.getPointInTime("user-1", 300)).toBe(3.0); // inclusive of event time
    expect(s.getPointInTime("user-1", 100)).toBe(1.0);
  });

  it("returns null before the first value exists", () => {
    expect(store().getPointInTime("user-1", 50)).toBeNull();
  });

  it("never leaks a value recorded after the event", () => {
    // At event 150 only the ts=100 value existed; ts=200/300 are the future.
    expect(store().getPointInTime("user-1", 150)).toBe(1.0);
  });

  it("expires features older than max staleness", () => {
    const s = store();
    expect(s.getPointInTime("user-1", 1000, 100)).toBeNull();
    expect(s.getPointInTime("user-1", 350, 100)).toBe(3.0);
  });

  it("treats the max-staleness boundary as inclusive", () => {
    const s = store();
    // event 400, ts=300 -> exactly 100 old; 100 is not "> 100", so kept.
    expect(s.getPointInTime("user-1", 400, 100)).toBe(3.0);
    // One unit older tips past the boundary and expires it.
    expect(s.getPointInTime("user-1", 401, 100)).toBeNull();
  });

  it("returns null for an unknown entity", () => {
    const s = store();
    expect(s.getPointInTime("ghost", 200)).toBeNull();
    expect(s.getOnline("ghost")).toBeNull();
  });
});

describe("FeatureStore training set", () => {
  it("is point-in-time correct across a spine", () => {
    const s = store();
    const spine = [
      { entity: "user-1", eventTime: 120 },
      { entity: "user-1", eventTime: 220 },
      { entity: "user-1", eventTime: 320 },
    ];
    expect(s.getTrainingSet(spine)).toEqual([1.0, 2.0, 3.0]);
  });

  it("leaves rows with no available value as null", () => {
    const s = store();
    const spine = [
      { entity: "user-1", eventTime: 50 },
      { entity: "user-1", eventTime: 250 },
      { entity: "ghost", eventTime: 999 },
    ];
    expect(s.getTrainingSet(spine)).toEqual([null, 2.0, null]);
  });

  it("returns an empty column for an empty spine", () => {
    expect(store().getTrainingSet([])).toEqual([]);
  });
});

describe("FeatureStore online lookup", () => {
  it("returns the latest value for an entity", () => {
    expect(store().getOnline("user-1")).toBe(3.0);
  });
});

describe("FeatureStore ordering", () => {
  it("handles out-of-order ingest", () => {
    const s = new FeatureStore();
    for (const [t, v] of [
      [300, 3.0],
      [100, 1.0],
      [200, 2.0],
    ] as const) {
      s.ingest({ entity: "e", timestamp: t, value: v });
    }
    expect(s.getPointInTime("e", 250)).toBe(2.0);
    expect(s.getOnline("e")).toBe(3.0);
  });

  it("serves the largest value when timestamps tie", () => {
    // The documented tie-break: values are ordered by (timestamp, value), so
    // among equal timestamps the largest value is served.
    const s = new FeatureStore();
    for (const [t, v] of [
      [200, 5.0],
      [200, 2.0],
      [200, 9.0],
      [100, 1.0],
    ] as const) {
      s.ingest({ entity: "e", timestamp: t, value: v });
    }
    expect(s.getPointInTime("e", 250)).toBe(9.0);
    expect(s.getOnline("e")).toBe(9.0);
    expect(s.getPointInTime("e", 150)).toBe(1.0);
  });
});

describe("FeatureStore edge cases", () => {
  it("returns null everywhere on an empty store", () => {
    const s = new FeatureStore();
    expect(s.getPointInTime("nobody", 10)).toBeNull();
    expect(s.getOnline("nobody")).toBeNull();
  });

  it("respects single-observation boundaries", () => {
    const s = new FeatureStore();
    s.ingest({ entity: "e", timestamp: 100, value: 1.0 });
    expect(s.getPointInTime("e", 99)).toBeNull(); // strictly before -> nothing
    expect(s.getPointInTime("e", 100)).toBe(1.0);
    expect(s.getPointInTime("e", 101)).toBe(1.0);
  });

  it("serves negative and zero values", () => {
    const s = new FeatureStore();
    s.ingest({ entity: "e", timestamp: 10, value: 0.0 });
    s.ingest({ entity: "e", timestamp: 20, value: -3.5 });
    expect(s.getPointInTime("e", 15)).toBe(0.0);
    expect(s.getPointInTime("e", 25)).toBe(-3.5);
    expect(s.getOnline("e")).toBe(-3.5);
  });
});
